Add initial CI pipeline with UI tests and axe scanning

chingucoding opened the pull request:
This PR adds the initial CI pipeline with azure devops, re-enable the UI tests and add Axe Accessibility scanning.

Closes #1024

kmahone commented:
We should avoid adding a .pfx key to the repo. This can cause issues, since they expire and also get flagged by scanning tools as being 'secrets' even in cases such as this where it is just for testing.

In WinUI, we use a script to generate the .pfx key at build time  [GenerateTestPfx.ps1](https://github.com/microsoft/microsoft-ui-xaml/blob/main/GenerateTestPfx.ps1)
A similar approach might work well here.

kmahone commented on WinUIGallery/Package.WAP.Dev.appxmanifest:
```diff
@@ -7,7 +7,7 @@
     IgnorableNamespaces="uap mp uap3">
   <!--Note: package version and assembly version must be kept in sync-->
-  <Identity Name="Microsoft.WinUI3ControlsGallery.Debug" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" Version="2.1.0.0" />
+  <Identity Name="Microsoft.WinUI3ControlsGallery.Debug" Publisher="CN=Microsoft Corporation" Version="2.1.0.0" />
```
>Publisher="CN=Microsoft Corporation" [](http://example.com/codeflow?start=57&length=36)

If this is being used for testing purposes, we could update the Publisher to something like "CN=WinUI-Gallery-Test" so it is clear that it is not an official build of something from Microsoft.

kmahone commented on UITests/TestBase.cs:
```diff
@@ -0,0 +1,50 @@
+			search.SendKeys(name);
+			Thread.Sleep(1_000);
+			Session.FindElementByName(name).Click();
+			Thread.Sleep(5_000);
```
Can we do something better than thread sleeps here? This seems unreliable.

E.g. can we wait for a certain piece of UI to update instead?

chingucoding replied:
I'm not really sure how you would do that with selenium, if you have any pointers I would love to do that.

chingucoding replied:
I also thought about that API, but its quite sophisticated. IIRC, part of that infra was also moved into a package right? Or was that something else.

Regarding the polling approach, I'm not sure if there is something sensible we want to wait for. The idea was to wait for the app to navigate to the page but I don't think we can "poll" for this without introducing UIA elements.

kmahone replied:
In the case of a page navigate you could wait until a known element on the page is available in the UIA tree.

kmahone commented:
> global using OpenQA.Selenium;

I don't see the advantage of doing this. It makes more sense to include the using directives in the files that require them.

---
Refers to: UITests/GlobalUsings.cs:1 in 07c17dd. [](commit_id = 07c17dd72cc19bf9d05796e75f5d9e6a14e767b9, deletion_comment = False)

kmahone commented on UITests/TestBase.cs:
```diff
@@ -0,0 +1,50 @@
+			Thread.Sleep(1_000);
+			Session.FindElementByName(name).Click();
+			Thread.Sleep(5_000);
+			AxeHelper.AssertNoAccessibilityErrors();
```
We should add explicit Axe test methods, instead of having this validation piggy-back on other tests via this helper method.

kmahone replied:
E.g. Tests\Button.cs should be updated to have
```
[TestMethod]
public void ValidateAccessibilityWithAxe()
```
or something like that.

kmahone commented on UITests/TestBase.cs:
```diff
@@ -0,0 +1,50 @@
+
+		[ClassCleanup]
+		public static void Cleanup()
+		{
```
What is the purpose of this Cleanup step?

kmahone reviewed (changes requested):
:clock1:

chingucoding commented:
I think I addressed most of your feedback @kmahone, the global usings file is gone. accessibility testing is an explicit test now. Regarding waiting for an element, I would appreciate some pointers on how to set this up.

kmahone commented:
> I think I addressed most of your feedback @kmahone, the global usings file is gone. accessibility testing is an explicit test now. Regarding waiting for an element, I would appreciate some pointers on how to set this up.

Thanks. I've updated the comment with a suggested approach for this.

kmahone commented on UITests/TestBase.cs:
```diff
@@ -0,0 +1,73 @@
+					break;
+				}
+				Thread.Sleep(50);
+			}
```
>} [](http://example.com/codeflow?start=3&length=1)

We should throw in the case where we never find the element.

chingucoding replied:
Gread idea, fixed that.

kmahone commented:
Looks good to me. @karkarl should take a look and approve also before merge.

chingucoding commented:
After/before merging, of course some Azure DevOps project also needs to be configured which is not something that I can do.

kmahone commented:
> After/before merging, of course some Azure DevOps project also needs to be configured which is not something that I can do.

Yes. Karen is going to set that up.
